src/Map.ts

```diff
@@ -0,0 +1,160 @@
+import { hasClass } from './decorators';
+import { is } from './object';
+import global from './global';
+
+export default class Map<K, V> {
+	private _keys: K[] = [];
+	private _values: V[] = [];
+
+	/*
+	 * An alternative to Array.prototype.indexOf using Object.is
+	 * to check for equality. See http://mzl.la/1zuKO2V
+	 */
+	private _indexOfKey(keys: K[], key: K): number {
+		for (var i = 0; i < keys.length; i++) {
+			if (is(keys[i], key)) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	/**
+	 * Creates a new Map
+	 *
+	 * @constructor
+	 *
+	 * @param iterable
+	 * An array of two-item tuples used to initially
+	 * populate the map. The first item in each tuple
+	 * corresponds to the key of the map entry. The second
+	 * item corresponds to the value of the map entry.
+	 */
+	constructor(iterable?: any) {
+		if (!(this instanceof Map)) {
+			throw new TypeError('Constructor Map requires "new"');
+		}
+
+		if (iterable) {
+			for (let [key, value] of iterable) {
+				this.set(key, value);
+			}
+		}
+	}
+
+	/**
+	 * Returns the number of key / value pairs in the Map
+	 *
+	 * @return the number of key / value pairs in the Map
+	 */
+	get size(): number {
+		return this._keys.length;
+	}
+
+	/**
+	 * Deletes all keys and their associated values.
+	 */
+	clear(): void {
+		this._keys.length = this._values.length = 0;
+	}
+
+	/**
+	 * Deletes a given key and its associated value.
+	 *
+	 * @param key The key to delete
+	 * @return true if the key exists, false if it does not
+	 */
+	delete(key: K): boolean {
+		const index = this._indexOfKey(this._keys, key);
+		if (index < 0) {
+			return false;
+		}
+		this._keys.splice(index, 1);
+		this._values.splice(index, 1);
+		return true;
+	}
+
+	/**
+	 * Returns an array of two-value tuples in the form
+	 * of [key, value] in order of insertion.
+	 *
+	 * @return an array of entries in order of insertion
+	 */
+	entries(): any[] {
+		var entries: any[] = [];
+		this._keys.forEach(function(key: K, index: number) {
+			entries.push([key, this.get(key)]);
```

Comment:
You added the `index` parameter but didn't actually use it here like I suggested.  `this.get(key)` becomes `this._values[index]` to skip an extra invocation.

I'm considering switching this to a straight-up for loop, while I'm at it, actually.